distributed/worker.py

```diff
@@ -1545,18 +1539,22 @@ def transition_dep_flight_waiting(self, dep, worker=None, remove=True):
                 pdb.set_trace()
             raise
 
-    def transition_dep_flight_memory(self, dep, value=None):
+    def transition_flight_memory(self, ts, value=None):
         try:
             if self.validate:
-                assert dep in self.in_flight_tasks
+                assert ts.key in self.in_flight_tasks
 
-            del self.in_flight_tasks[dep]
-            if self.dependents[dep]:
-                self.dep_state[dep] = "memory"
-                self.put_key_in_memory(dep, value)
-                self.batched_stream.send({"op": "add-keys", "keys": [dep]})
+            del self.in_flight_tasks[ts.key]
+            if ts.dependents:
+                self.put_key_in_memory(ts, value)
+                assert ts.key in self.data
```

Comment:
If the value is large enough to be placed into slow storage immediately (unlikely) then this might not be free.  We might want to drop it.